### PR TITLE
feat: add map step type in sql

### PR DESCRIPTION
**Pull request by jumski**
# Map Step Type Implementation (SQL Core)

## Overview

This PR implements **map step type** functionality in pgflow's SQL Core, enabling parallel processing of array data through fanout/map operations. This foundational feature allows workflows to spawn multiple tasks from a single array input, process them in parallel, and aggregate results back into ordered arrays.

### Core Concept: Map Steps

Map steps transform single array inputs into parallel task execution:

```typescript
// Input array
const items = ['item1', 'item2', 'item3'];

// Map step spawns 3 parallel tasks
// Each task processes one array element
// Results aggregated back: ["result1", "result2", "result3"]
```

### Two Types of Map Steps

**1. Root Maps (0 dependencies)**

- Input: `runs.input` must be a JSONB array
- Behavior: Spawns N tasks where N = array length
- Example: `start_flow('my_flow', ["a", "b", "c"])` → 3 parallel tasks

**2. Dependent Maps (1 dependency)**

- Input: Output from a completed dependency step (must be array)
- Behavior: Spawns N tasks based on dependency's output array length
- Example: `single_step → ["x", "y"]` → `map_step` spawns 2 tasks

### Map→Map Chaining

Maps can depend on other maps:

- Parent map completes all tasks first
- Child map receives parent's task count as its planned count
- Child tasks read from aggregated parent array using simple approach

### Hybrid Architecture: "Fresh Data" vs "Dumb Spawning"

**Key Innovation**: Separate array parsing from task spawning for optimal performance.

**Fresh Data Functions** (parse arrays, set counts):

- `start_flow()`: Validates `runs.input` arrays for root maps, sets `initial_tasks`
- `complete_task()`: Validates dependency outputs for dependent maps, sets `initial_tasks`

**Dumb Functions** (use pre-computed counts):

- `start_ready_steps()`: Copies `initial_tasks → remaining_tasks` and spawns N tasks (no JSON parsing)
- `start_tasks()`: Extracts array elements using `task_index`
- `maybe_complete_run()`: Aggregates results into ordered arrays

### Schema Changes

1. **Enable Map Step Type**

   ```sql
   -- steps table constraint updated
   check (step_type in ('single', 'map'))
   ```

2. **Remove Single Task Limit**

   ```sql
   -- step_tasks table constraint removed
   -- constraint only_single_task_per_step check (task_index = 0)
   ```

3. **Add Task Planning Columns**

   ```sql
   -- step_states table updates
   remaining_tasks int NULL,  -- NULL = not started, >0 = active countdown
   initial_tasks int DEFAULT 1 CHECK (initial_tasks >= 0),  -- Planned count
   CONSTRAINT remaining_tasks_state_consistency CHECK (
     remaining_tasks IS NULL OR status != 'created'
   )
   ```

4. **Enhanced Functions**
   - `add_step()`: Accepts `step_type` parameter, validates map constraints
   - `start_flow()`: Root map validation and count setting
   - `complete_task()`: Dependent map count setting
   - `start_ready_steps()`: Efficient bulk task spawning
   - `start_tasks()`: Array element extraction
   - `maybe_complete_run()`: Result aggregation

### Performance Optimizations

- **Minimal JSON Parsing**: Arrays parsed exactly twice (once in each "fresh data" function)
- **Batch Operations**: Efficient task creation using `generate_series(0, N-1)`
- **Atomic Updates**: Count setting under existing locks

### Edge Cases Handled

- **Empty Arrays**: Auto-complete with `[]` output (not failures)
- **Array Validation**: Fail fast with clear errors for non-array inputs
- **Mixed Workflows**: Single + map steps in same flow
- **Task Indexing**: 0-based indexing matching JSONB array indices
- **Ordered Results**: Results aggregated in `task_index` order

### Root Map Flow

```sql
-- 1. Create flow with root map step
SELECT pgflow.create_flow('example_flow');
SELECT pgflow.add_step('example_flow', 'process_items', 'map');

-- 2. Start with array input
SELECT pgflow.start_flow('example_flow', '["item1", "item2", "item3"]');
-- → Validates input is array
-- → Sets initial_tasks = 3 for 'process_items' step

-- 3. start_ready_steps() spawns 3 tasks
-- → Creates step_tasks with task_index 0, 1, 2
-- → Sends 3 messages to pgmq queue

-- 4. Workers process tasks in parallel
-- → Each gets individual array element: "item1", "item2", "item3"
-- → Returns processed results

-- 5. Results aggregated in run output
-- → Final output: {"process_items": ["result1", "result2", "result3"]}
```

### Dependent Map Flow

```sql
-- 1. Single step produces array output
SELECT pgflow.complete_task(run_id, 'prep_step', 0, '["a", "b"]');
-- → complete_task() detects map dependent
-- → Sets initial_tasks = 2 for dependent map step

-- 2. Map step becomes ready and spawns 2 tasks
-- → Each task gets individual element: "a", "b"
-- → Processes in parallel

-- 3. Results aggregated
-- → Map step output: ["processed_a", "processed_b"]
```

## Handler Interface

Map task handlers receive individual array elements (TypeScript `Array.map()` style):

```typescript
// Single step handler
const singleHandler = (input: {run: Json, dep1?: Json}) => { ... }

// Map step handler
const mapHandler = (item: Json) => {
  // Receives just the array element, not wrapped object
  return processItem(item);
}
```

## Comprehensive Testing

**33 test files** covering:

- Schema validation and constraints
- Function-by-function unit tests
- Integration workflows (root maps, dependent maps, mixed)
- Edge cases (empty arrays, validation failures)
- Error handling and failure propagation
- Idempotency and double-execution safety

## Benefits

- **Parallel Processing**: Automatic fanout for array data
- **Type Safety**: Strong validation ensures arrays where expected
- **Performance**: Minimal JSON parsing, efficient batch operations
- **Simplicity**: Familiar map semantics, clear error messages
- **Scalability**: Handles large arrays efficiently
- **Composability**: Map steps integrate seamlessly with existing single steps

## Breaking Changes

None. This is a pure addition:

- Existing single steps unchanged
- New `step_type` parameter defaults to `'single'`
- Backward compatibility maintained
- Database migrations are additive only

## Implementation Status

- ✅ **Architecture Designed**: Hybrid approach validated
- ✅ **Plan Documented**: Complete implementation roadmap
- 🔄 **In Progress**: Schema changes and function updates
- ⏳ **Next**: TDD implementation starting with `add_step()` function

### CODE REVIEW

**Comment by jumski** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/208?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#231** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/231?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#230** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#228** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/228?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#222** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/222?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#221** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/221?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#220** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/220?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#219** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/219?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#218** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/218?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#217** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/217?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#216** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#215** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/215?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#214** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/214?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#213** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/213?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#209** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#208** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/208?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/208?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#207** <a href="https://app.graphite.dev/github/pr/pgflow-dev/pgflow/207?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>merge:queue</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix:queue</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/pgflow-dev?ref=merge-queue-instructions-in-stack-comment&prId=208).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: pkgs/core/schemas/0100_function_add_step.sql [hunk 1]
```diff
@@ -1,56 +1,94 @@
...
-  max_attempts int default null,
-  base_delay int default null,
-  timeout int default null,
-  start_delay int default null
+  step_type text default 'single',
+  deps_slugs text [] default '{}',
+  opt_max_attempts int default null,
+  opt_base_delay int default null,
+  opt_timeout int default null,
+  opt_start_delay int default null
```
Comment (jumski) [inline pkgs/core/schemas/0100_function_add_step.sql]: we dont want to rename parameters

File: pkgs/core/schemas/0100_function_add_step.sql [hunk 2]
```diff
@@ -1,56 +1,94 @@
...
+    COALESCE($3, 'single'), -- step_type
+    next_idx, 
+    COALESCE(array_length($4, 1), 0), -- deps_slugs
+    $5, -- opt_max_attempts
+    $6, -- opt_base_delay
+    $7, -- opt_timeout
+    $8  -- opt_start_delay
+  )
+  ON CONFLICT ON CONSTRAINT steps_pkey
+  DO UPDATE SET step_slug = EXCLUDED.step_slug
```
[the remark this answers is not in the document]
Comment (jumski) [reply pkgs/core/schemas/0100_function_add_step.sql:53]: we do a no-op update to be able to do RETURNING *, otherwise its impossible right?

File: pkgs/core/supabase/tests/add_step/simple_map_test.test.sql [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+-- Test 2: Check step_type
+SELECT is(
+  (SELECT step_type FROM pgflow.steps WHERE flow_slug = 'test_flow' AND step_slug = 'map_step'),
+  'map',
+  'Map step should have step_type = "map"'
+);
+
+SELECT * FROM finish();
+
+ROLLBACK;
```
Comment (jumski) [inline pkgs/core/supabase/tests/add_step/simple_map_test.test.sql]: i would remove this whole test file altogether, it is repetitive and does not bring anything to the table

File: pkgs/core/supabase/tests/add_step/step_type_validation.test.sql [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+-- Test 1: Invalid step_type should be rejected - use throws_like to match constraint error
+SELECT throws_like($$
+  SELECT pgflow.add_step(
+    flow_slug => 'test_flow',
+    step_slug => 'invalid_step',
+    step_type => 'invalid_type',
+    deps_slugs => '{}'
+  );
+$$, '%check constraint%', 'Should reject invalid step_type');
```
Comment (jumski) [inline pkgs/core/supabase/tests/add_step/step_type_validation.test.sql]: the assertion on error message should be more specific because it can fail some other check

File: pkgs/core/supabase/tests/add_step/step_type_validation.test.sql [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+-- Test 2: NULL step_type should default to 'single'
+SELECT lives_ok($$
+  SELECT pgflow.add_step(
+    flow_slug => 'test_flow',
+    step_slug => 'null_type_step',
+    step_type => NULL,
+    deps_slugs => '{}'
+  );
+$$, 'Should accept NULL step_type and default to single');
```
Comment (jumski) [inline pkgs/core/supabase/tests/add_step/step_type_validation.test.sql]: this is tested elsewhere
